resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -200,58 +224,94 @@ private[spark] class KubernetesClusterSchedulerBackend(
           .withContainerPort(port._2)
           .build()
       })
-    try {
-      (executorId, kubernetesClient.pods().createNew()
-        .withNewMetadata()
-          .withName(name)
-          .withLabels(selectors)
-          .withOwnerReferences()
-          .addNewOwnerReference()
-            .withController(true)
-            .withApiVersion(driverPod.getApiVersion)
-            .withKind(driverPod.getKind)
-            .withName(driverPod.getMetadata.getName)
-            .withUid(driverPod.getMetadata.getUid)
-          .endOwnerReference()
+
+    val pod = new PodBuilder()
+      .withNewMetadata()
+        .withName(name)
+        .withLabels(selectors)
+        .withOwnerReferences()
+        .addNewOwnerReference()
+          .withController(true)
+          .withApiVersion(driverPod.getApiVersion)
+          .withKind(driverPod.getKind)
+          .withName(driverPod.getMetadata.getName)
+          .withUid(driverPod.getMetadata.getUid)
+        .endOwnerReference()
         .endMetadata()
-        .withNewSpec()
-          .withHostname(hostname)
-          .addNewContainer()
-            .withName(s"executor")
-            .withImage(executorDockerImage)
-            .withImagePullPolicy("IfNotPresent")
-            .withNewResources()
-              .addToRequests("memory", executorMemoryQuantity)
-              .addToLimits("memory", executorMemoryLimitQuantity)
-              .addToRequests("cpu", executorCpuQuantity)
-              .addToLimits("cpu", executorCpuQuantity)
-              .endResources()
-            .withEnv(requiredEnv.asJava)
-            .withPorts(requiredPorts.asJava)
-            .endContainer()
-          .endSpec()
-        .done())
+      .withNewSpec()
+        .withHostname(hostname)
+        .addNewContainer()
+          .withName(s"executor")
+          .withImage(executorDockerImage)
+          .withImagePullPolicy("IfNotPresent")
+          .withNewResources()
+            .addToRequests("memory", executorMemoryQuantity)
+            .addToLimits("memory", executorMemoryLimitQuantity)
+            .addToRequests("cpu", executorCpuQuantity)
+            .addToLimits("cpu", executorCpuQuantity)
+          .endResources()
+          .withEnv(requiredEnv.asJava)
+          .withPorts(requiredPorts.asJava)
+        .endContainer()
+      .endSpec()
+      .build()
+
+      shuffleDir match {
```

Comment:
Use `Option.map` instead of `match` here. Using `map` and `getOrElse` we can use a more fluent builder pattern. Perhaps something like this:
    
    val basePodBuilder = new PodBuilder().<default configuration>.endSpec()
    val resolvedPodBuilder = shuffleDir.map { dir =>
      basePodBuilder.editSpec()
        .setVolumes(..)
        .editContainerAt(0).<edits>.endContainer() // Can also use editMatchingContainer()
        .endSpec()
    }.getOrElse(basePodBuilder)
    kubernetesClient.pods().create(resolvedPod.build())

We could also build the container separately from the pod, but I'm not sure how much that buys us. That would look something like this:

    val baseContainerBuilder = new ContainerBuilder().<default configurations>
    val basePodBuilder = new PodBuilder().<default configuration>.endSpec()
    val resolvedPodBuilder = shuffleDir.map { dir =>
       basePodBuilder.editSpec()
         .setVolumes(..)
         .endSpec()
     }.getOrElse(basePodBuilder)
     val resolvedContainerBuilder = shuffleDir.map { dir =>
       baseContainerBuilder.addToVolumeMounts(...)
     }.getOrElse(baseContainerBuilder)
     val resolvedPod = resolvedPodBuilder().editSpec()
       .addContainer(resolvedContainerBuilder.build())
       .endSpec()
       .build()